Refactor Roving Tab Index Implementations

Westbrook opened the pull request:
## Description
- create the `@spectrum-web-components/reactive-controllers` package. Does that sound OK?
- vend a Roving Tab Index Reactive Controller
- refactor Roving Tab Index usage to leverage the Reactive Controller
- remove `PageUp` and `PageDown` functionality from `sp-action-group` and `sp-radio-group` as it's not actually in the w3c or Spectrum specs for those patterns.

## Related issue(s)

- fixes #1999

## Motivation and context
Centralize, abstract, make less.

## How has this been tested?

-   [ ] _Test case 1_
    1. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/accordion--default)
    2. Click an Accordion Item
    3. Use `ArrowUp`/`ArrowDown` to navigate to another Item and `Enter`/`Space` to toggle that item
    4. Use `Home`/`End` to jump to the first/last Item
-   [ ] _Test case 2_
    1. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/action-group--default)
    2. Click an Action Button
    3. Use `ArrowUp`/`ArrowDown`/`ArrowLeft`/`ArrowRight` to navigate to another Button
    4. Use `Home`/`End` to jump to the first/last Button
-   [ ] _Test case 3_
    1. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/radio--group-example)
    2. Click an Radio Button
    3. Use `ArrowUp`/`ArrowDown`/`ArrowLeft`/`ArrowRight` to navigate to AND select another Button
    4. Use `Home`/`End` to jump to AND select the first/last Button
-   [ ] _Test case 4_
    1. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/sidenav--default): unmanaged tab index
    2. Click a Sidenav Item
    3. Use `Tab` to navigate through the various Sidenav Items
    4. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/sidenav--manage-tab-index): managed index
    5. Click a Sidenav Item
    6. Use `ArrowUp`/`ArrowDown` to navigate to another Sidenav Item and `Enter` to toggle that item
    7. Use `Home`/`End` to jump to the first/last Item
-   [ ] _Test case 5_
    1. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/tabs--default): horizontal
    2. Click a Tab Item
    3. Use `ArrowLeft`/`ArrowRight` to navigate to another Tab and `Enter`/`Space` to toggle that item
    4. Use `Home`/`End` to jump to the first/last Item
    5. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/tabs--vertical): vertical
    6. Use `ArrowUp`/`ArrowDown` to navigate to another Tab and `Enter`/`Space` to toggle that item
    7. Use `Home`/`End` to jump to the first/last Item
-   [ ] _Test case 6_
    1. Go [here](https://roving-tab-index--spectrum-web-components.netlify.app/storybook/index.html?path=/story/tags--deletable)
    2. Click a Tag Item
    3. Use `ArrowUp`/`ArrowDown`/`ArrowLeft`/`ArrowRight` to navigate to another Tag and `Delete`/`Backspace`/`Space` to "delete" that item, show in the Actions pane of Storybook.
    4. Use `Home`/`End` to jump to the first/last Item
-   [ ] _Test case 7_
    1. [x] Integration testing required!

## Types of changes
-   [x] New feature (non-breaking change which adds functionality)

## Checklist
-   [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
-   [x] My code follows the code style of this project.
-   [x] If my change required a change to the documentation, I have updated the documentation in this pull request.
-   [x] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** document.
-   [x] I have added tests to cover my changes.
-   [x] All new and existing tests passed.

hunterloftis commented:
In test case 2, is this expected behavior? It seems like the first element may be being treated differently from the other elements.

1. Click on middle button
2. Press [right arrow]
3. Focus moves to right button
4. Click on left button
5. Press [right arrow]
6. Focus moves to left button
7. (repeat with [down arrow])

https://user-images.githubusercontent.com/364501/150224302-35081e88-617f-4140-845d-1270b067f1b3.mov

hunterloftis commented:
In test case 5, arrow keys & home/end don't produce any visible change for me (neither does space). Tab & enter works as expected though:

https://user-images.githubusercontent.com/364501/150229446-d7ee8bfa-4f90-45db-ae7d-3e7eadf01637.mov

hunterloftis commented:
Test case 6 works as expected for tags, although space/enter doesn't toggle anything (I wouldn't expect it to - just making sure I understand the test instructions)

Westbrook commented:
@hunterloftis thanks for getting started here, updates as follows.

Test case 2:
- corrected  multiple uses of `0 || other` fall through in logic to `??` or other more reliable logic
- allowed `focusin` based index management while focus is held

Test case 4:
- linked to the wrong story; Sidenav has both managed and unmanaged versions
- prevented the unamanaged version from accidentally becoming managed so that all elements are `tabindex=0` to start

Test case 6:
- added `delete` action to Storybook so that the `Delete`/`Backspace`/`Space`  key will provide visible reaction

Westbrook commented:
Writing test cases in English is a very strange experience 🙃 I've expanded that Tabs context to include a link to the _vertical_ instance as well. Horizontal should accept `ArrowLeft` and `ArrowRight` while vertical accepts `ArrowUp` and `ArrowDown`...

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        );
+    }
+
+    setCurrentIndexCircularly(diff: number): void {
+        const { length } = this.elements;
+        let nextIndex = (length + this.currentIndex + diff) % length;
+        while (
+            this.elements[nextIndex] &&
+            !this.isFocusableElement(this.elements[nextIndex])
+        ) {
+            nextIndex = (length + nextIndex + diff) % length;
+        }
+        this.currentIndex = nextIndex;
+    }
```
`while`s always make me start thinking defensively about the potential for infinite loops.

What it ever be possible to enter this function with 10 elements that test negative for `isFocusableElement` (either by being disabled, or by the user providing a bad `isFocusableElement` like `return false`)?

To be extra safe, I might limit iterations to `length`, since we never want to loop in more than a circle:

```ts
for (let i = 0; i < length; i++) {
  const index = (length + i + diff) % length;
  if (this.elements[index] && this.isFocusableElement(this.elements[index])) {
    this.currentIndex = index;
    return;
  }
}
```

Less dangerous: could length ever be 0? (x % 0 = NaN)

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        this.host.addEventListener('keydown', this.handleKeydown);
+        this.focused = true;
+        const currentElement = this.elements[this.currentIndex];
+        if (!currentElement) return;
```
What is the purpose of this if statement?

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        this.currentIndex = this.focusInIndex;
+        this.focused = false;
+        const currentElement = this.elements[this.currentIndex];
+        if (!currentElement) return;
```
Ditto here

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        }
+        const path = event.composedPath() as T[];
+        let targetIndex: number | undefined;
+        path.find((el) => {
```
maybe another opportunity for `.findIndex`

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+            targetIndex = this.elements.indexOf(el);
+            return targetIndex !== -1;
+        });
+        this.currentIndex = targetIndex ?? this.currentIndex;
```
This may be doing exactly what you want as right now `targetIndex` can be `undefined`, but since indices are typically invalid at `-1` and valid at `0`, I'd recommend:

```suggestion
        this.currentIndex = targetIndex > -1 ? targetIndex : this.currentIndex;
```

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        }
+    };
+
+    acceptsEventCode(code: string): boolean {
+        if (code === 'End' || code === 'Home') {
+            return true;
+        }
+        switch (this.direction) {
+            case 'horizontal':
+                return code === 'ArrowLeft' || code === 'ArrowRight';
+            case 'vertical':
+                return code === 'ArrowUp' || code === 'ArrowDown';
+            case 'both':
+            case 'grid':
+                return code.startsWith('Arrow');
+        }
+    }
```
I am guaranteed to typo these strings someday if we don't make them constants 😅

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+            return;
+        }
+        let diff = 0;
+        switch (event.code) {
+            case 'ArrowRight':
+                diff += 1;
+                event.preventDefault();
+                break;
+            case 'ArrowDown':
+                diff += this.direction === 'grid' ? this.directionLength : 1;
+                event.preventDefault();
+                break;
+            case 'ArrowLeft':
+                diff -= 1;
+                event.preventDefault();
+                break;
+            case 'ArrowUp':
+                diff -= this.direction === 'grid' ? this.directionLength : 1;
+                event.preventDefault();
+                break;
+            case 'End':
+                this.currentIndex = 0;
+                diff -= 1;
+                event.preventDefault();
+                break;
+            case 'Home':
+                this.currentIndex = this.elements.length - 1;
+                diff += 1;
+                event.preventDefault();
+                break;
+        }
```
Would this accomplish the same thing, or is there a reason not to bail if none of our codes match?

```suggestion
        switch (event.code) {
            case 'ArrowRight':
                diff += 1;
                break;
            case 'ArrowDown':
                diff += this.direction === 'grid' ? this.directionLength : 1;
                break;
            case 'ArrowLeft':
                diff -= 1;
                break;
            case 'ArrowUp':
                diff -= this.direction === 'grid' ? this.directionLength : 1;
                break;
            case 'End':
                this.currentIndex = 0;
                diff -= 1;
                break;
            case 'Home':
                this.currentIndex = this.elements.length - 1;
                diff += 1;
                break;
            default:
                return;
        }
        event.preventDefault();
```

hunterloftis commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        if (this.focused) {
+            this.updateTabindexes(() => -1);
+        } else {
+            this.updateTabindexes((el: HTMLElement): number | false => {
+                const containsFocusInElement = el.contains(this.focusInElement);
+                if (el === this.focusInElement) {
+                    return 0;
+                } else if (containsFocusInElement) {
+                    return false;
+                }
+                return -1;
+            });
```
This is a weird function signature (0, -1, or `false`). I don't think this is necessarily better, but wanted to think of an alternative:

```suggestion
            this.updateTabindexes((el: HTMLElement): number | undefined => {
                if (el === this.focusInElement) return 0;
                return el.contains(this.focusInElement) ? undefined : -1;
            });
```

Not really an improvement to `updateTabindexes` either; maybe we can simplify these later instead:

```ts
    updateTabindexes(getTabIndex: (el: HTMLElement) => number | undefined): void {
        this.elements.forEach((el) => {
            const tabIndex = getTabIndex(el);
            if ([0, -1].includes(tabIndex)) {
                el.tabIndex = tabIndex;
                return;
            }
            el.removeAttribute('tabindex');
            const updatable = el as unknown as {
                requestUpdate?: () => void;
            };
            if (updatable.requestUpdate) updatable.requestUpdate();
        });
    }
```

hunterloftis reviewed (approved):
![image](https://user-images.githubusercontent.com/364501/150392917-726b8141-0684-4da2-aaec-4cfe19ed0f5b.png)

I left a bunch of feedback to consider, but all the tests work as described now & there's nothing blocking here.

Westbrook commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        this.host.addController(this);
+        this._elements = elements;
+        this.isFocusableElement = isFocusableElement || this.isFocusableElement;
+        if (typeof direction === 'string') {
+            this._direction = () => direction;
+        } else if (typeof direction === 'function') {
+            this._direction = direction;
+        }
```
`direction` isn't required in the `constructor` and defaults to `_direction` defaults to `() => 'both'` in the class.

Westbrook commented on tools/reactive-controllers/src/RovingTabindex.ts:
```diff
@@ -0,0 +1,304 @@
+        }
+        this.calculateFocusInIndex =
+            calculateFocusInIndex || this.calculateFocusInIndex;
+        this.elementEnterAction = elementEnterAction || this.elementEnterAction;
```
This looks like it forces the `undefined` into this? Am I missing something?

hunterloftis replied:
🤦  yep you're right, I was missing something!

Westbrook commented on tools/reactive-controllers/roving-tab-index.md:
```diff
@@ -0,0 +1,92 @@
+A `Container` can further customize the implementation of the `RovingTabindexController` with the following options:
+
+-   `calculateFocusInIndex` to control what element will recieve `tabindex=0` while focus is outside of the `Container` and accepts a method with a signature of `(_elements: T[]) => number`
+-   `direction` to customize how and which arrow keys manage what element is to be focused and accepts a either a string of `both`, `vertical`, `horizontal`, or `grid` or a method returning one of those strings
```
Unimplemented..._but_ while "both" moves linear along the list regardless of input, "grid" will visually follow the layout of the elements, so `ArrowUp`/`ArrowDown` will "skip" the length of the row.

Westbrook reviewed (commented):
@hloftis made some changes on your great suggestions. Very useful! Let me know if this still looks good to you.
